plugins/janus_sipre.c

```diff
@@ -0,0 +1,2893 @@
+				if(!username)
+					g_snprintf(user_id, 255, "janus-sipre-%"SCNu32"", janus_random_uint32());
+				JANUS_LOG(LOG_INFO, "Guest will have username %s\n", user_id);
+				send_register = FALSE;
+			} else {
+				json_t *secret = json_object_get(root, "secret");
+				json_t *ha1_secret = json_object_get(root, "ha1_secret");
+				json_t *authuser = json_object_get(root, "authuser");
+				if(!secret && !ha1_secret) {
+					g_free(user_id);
+					g_free(user_host);
+					JANUS_LOG(LOG_ERR, "Missing element (secret or ha1_secret)\n");
+					error_code = JANUS_SIPRE_ERROR_MISSING_ELEMENT;
+					g_snprintf(error_cause, 512, "Missing element (secret or ha1_secret)");
+					goto error;
+				}
+				if(secret && ha1_secret) {
+					g_free(user_id);
+					g_free(user_host);
+					JANUS_LOG(LOG_ERR, "Conflicting elements specified (secret and ha1_secret)\n");
+					error_code = JANUS_SIPRE_ERROR_INVALID_ELEMENT;
+					g_snprintf(error_cause, 512, "Conflicting elements specified (secret and ha1_secret)");
+					goto error;
+				}
+				const char *secret_text;
+				if(secret) {
+					secret_text = json_string_value(secret);
+					session->account.secret = g_strdup(secret_text);
+					session->account.secret_type = janus_sipre_secret_type_plaintext;
+				} else {
+					secret_text = json_string_value(ha1_secret);
+					session->account.secret = g_strdup(secret_text);
+					session->account.secret_type = janus_sipre_secret_type_hashed;
+				}
+				if(authuser) {
+					const char *authuser_text;
+					authuser_text = json_string_value(authuser);
+					session->account.authuser = g_strdup(authuser_text);
+				} else {
+					session->account.authuser = g_strdup(user_id);
+				}
+				/* Got the values, try registering now */
+				JANUS_LOG(LOG_VERB, "Registering user %s (secret %s) @ %s through %s\n",
+					user_id, secret_text, user_host, proxy_text != NULL ? proxy_text : "(null)");
+			}
+
+			session->account.identity = g_strdup(username_text);
+			g_hash_table_insert(identities, session->account.identity, session);
+			session->account.sips = sips;
+			session->account.username = g_strdup(user_id);
+			if(display_name_text) {
+				session->account.display_name = g_strdup(display_name_text);
+			}
+			if(user_agent_text) {
+				session->account.user_agent = g_strdup(user_agent_text);
+			}
+			if(proxy_text) {
+				session->account.proxy = g_strdup(proxy_text);
+			} else {
+				/* Build one from the user's identity */
+				char uri[256];
+				g_snprintf(uri, sizeof(uri), "sip:%s:%"SCNu16, user_host, (user_port ? user_port : 5060));
+				session->account.proxy = g_strdup(uri);
+			}
+			g_free(user_host);
+			g_free(user_id);
+
+			session->account.registration_status = janus_sipre_registration_status_registering;
+			if(send_register) {
+				char ttl_text[20];
+				g_snprintf(ttl_text, sizeof(ttl_text), "%d", ttl);
+					/* TODO Any way to specify TTL in sipreg_register? */
+				int err = sipreg_register(&session->stack.reg, sipstack,
```

Comment:
Nope, this is called by the plugin message handler thread.

When you say the same thread that is running the re-main loop, do you mean the thread that calls `re_main(..)`? We have that call here:

https://github.com/meetecho/janus-gateway/blob/f0c374b524c514d175496040bc7017ffebb25333/plugins/janus_sipre.c#L2778

but my understanding was that the `re_main(..)` call is blocking, and so the thread can't do anything else, which is why I'm devoting a separate thread just for that. Is this the wrong way to use the method?